docs/components/security.rst

```diff
@@ -1,23 +1,365 @@
+    {
+        foreach ($permissions as $level => &$perms) {
+            foreach ($perms as $perm) {
+                $include = array();
+                switch ($perm) {
+                    case 'send_probe':
+                        $include = array('use_telescope');
+                        break;
+                    case 'visit':
+                        $include = array('use_telescope', 'send_probe');
+                        break;
+                }
+                if (!empty($include)) {
+                    foreach ($include as $r) {
+                        list($ignore, $r) = $this->getSynonym($level, $r);
+                        if ($this->isSupported($level, $r) && !in_array($r, $perms)) {
+                            $perms[] = $r;
+                        }
+                    }
+                }
+            }
+        }
+
+        // Return true if this method needs a second round after all the other bundles have adjusted permissions
+        return false;
+    }
+
+Plugins can adjust permissions based on other selected permissions to prevent ``User error``.
+For example, if a User has permission to ``edit``, then the User also needs permission to ``view`` whether that got selected in the Role Form or not.
+You can use the method ``analyzePermissions()`` for this which the Plugin the opportunity to modify permissions based on other selections before persisting to the database. 
+
+Sometimes, it may be necessary to re-adjust based on a permission that's outside the Plugin's control.
+In this case, ``analyzePermissions()`` can return true and it gets called again after all the permissions got analyzed by the other bundles and Plugins.
+In this case, the argument ``$isSecondRound`` is true.
+
+Advanced ``isGranted`` logic
+============================
+
+If it's necessary to perform some logic other than simply comparing bits, the permission class can override the parent's ``public function isGranted($userPermissions, $name, $level)`` and do whatever is necessary for it's own permission levels and individual permissions.
+
+Advanced ``isSupported`` logic
+==============================
+
+The same applies for the method ``isSupported()`` which you can use to determine if a bundle or Plugin includes the requested permission and permission level.
+You can also use this to provide BC support.
+
+.. vale off
 
 Single sign-on
 **************
 
+.. vale on
+
 Todo:
 
-\Mautic\UserBundle\UserEvents::USER_LOGIN
-\Mautic\UserBundle\UserEvents::USER_FORM_AUTHENTICATION
-\Mautic\UserBundle\UserEvents::USER_PRE_AUTHENTICATION
-\Mautic\UserBundle\UserEvents::USER_AUTHENTICATION_CONTENT
\ No newline at end of file
+- ``\Mautic\UserBundle\UserEvents::USER_LOGIN``
+- ``\Mautic\UserBundle\UserEvents::USER_FORM_AUTHENTICATION``
+- ``\Mautic\UserBundle\UserEvents::USER_PRE_AUTHENTICATION``
+- ``\Mautic\UserBundle\UserEvents::USER_AUTHENTICATION_CONTENT``
```

Comment:
I left this on `todo` because the current Developer Docs don't mention anything about this and I haven't seen any implementations of this yet. Happy to remove it as well so someone else who has experience with this can add it

Reply:
@escopecz something you could help with or shall we leave it as a todo and create an issue to update?